dask/dataframe/io/io.py

```diff
@@ -454,48 +454,68 @@ def from_dask_array(x, columns=None, index=None, meta=None):
-    to_merge = []
+    name = "from-dask-array-" + tokenize(x, columns)
+    graph_dependencies = [x]
+    arrays_and_indices = [x.name, "ij" if x.ndim == 2 else "i"]
+    numblocks = {x.name: x.numblocks}
 
     if index is not None:
-        if not isinstance(index, Index):
-            raise ValueError("'index' must be an instance of dask.dataframe.Index")
+        # An index is explicitly given by the caller, so we can pass it through to the
+        # initializer after a few checks.
         if index.npartitions != x.numblocks[0]:
             msg = (
                 "The index and array have different numbers of blocks. "
                 "({} != {})".format(index.npartitions, x.numblocks[0])
             )
             raise ValueError(msg)
         divisions = index.divisions
-        to_merge.append(ensure_dict(index.dask))
-        index = index.__dask_keys__()
+        graph_dependencies.append(index)
+        arrays_and_indices.extend([index._name, "i"])
+        numblocks[index._name] = (index.npartitions,)
 
     elif np.isnan(sum(x.shape)):
+        # The shape of the incoming array is not known in at least one dimension. As
+        # such, we can't create an index for the entire output DataFrame and we set
+        # the divisions to None to represent that.
         divisions = [None] * (len(x.chunks[0]) + 1)
-        index = [None] * len(x.chunks[0])
     else:
+        # The shape of the incoming array is known and we don't have an explicit index.
+        # Create a mapping of chunk number in the incoming array to
+        # (start row, stop row) tuples. These tuples will be used to create a sequential
+        # RangeIndex later on that is continuous over the whole DataFrame.
         divisions = [0]
-        for c in x.chunks[0]:
-            divisions.append(divisions[-1] + c)
-        index = [
-            (np.arange, a, b, 1, "i8") for a, b in zip(divisions[:-1], divisions[1:])
-        ]
+        stop = 0
+        index_mapping = {}
+        for i, increment in enumerate(x.chunks[0]):
+            stop += increment
+            index_mapping[(i,)] = (divisions[i], stop)
+            divisions.append(stop)
         divisions[-1] -= 1
+        arrays_and_indices.extend([BlockwiseDepDict(mapping=index_mapping), "i"])
 
-    dsk = {}
-    for i, (chunk, ind) in enumerate(zip(x.__dask_keys__(), index)):
-        if x.ndim == 2:
-            chunk = chunk[0]
-        if is_series_like(meta):
-            dsk[name, i] = (type(meta), chunk, ind, x.dtype, meta.name)
-        else:
-            dsk[name, i] = (type(meta), chunk, ind, meta.columns)
+    if is_series_like(meta):
+        kwargs = {"dtype": x.dtype, "name": meta.name}
+    else:
+        kwargs = {"columns": meta.columns}
+
+    def _partition_from_array(data, index=None, **kwargs):
+        if isinstance(index, tuple):
+            index = pd.RangeIndex(*index)
+        return type(meta)(data, index=index, **kwargs)
```

Comment:
It's my understanding that we usually try to avoid closures for tasks as they are more expensive to serialize. Instead, we could probably make this a top-level function and pass in `type(meta)` as an argument.

Reply:
>we usually try to avoid closures for tasks as they are more expensive to serialize

I'm guilty of doing this a lot, but I totally agree. Might as well define this as a top-level function to avoid serializing the entire `from_dask_array` code :)

Reply:
Done and done 😄